Lib/test/test_iter.py

```diff
@@ -267,6 +267,23 @@ def spam(state=[0]):
             return i
         self.check_iterator(iter(spam, 20), list(range(10)), pickle=False)
 
+    # Test two-argument iter() with function that empties its associated
+    # iterator via list() (or anything else but next()) but returns
+    # a non-sentinel value thus claiming that the iterator can yield more.
+    def test_iter_function_concealing_reentrant_exhaustion(self):
+        HAS_MORE = 1
+        NO_MORE = 2
+        def spam():
+            if spam.need_reentrance:
+                spam.need_reentrance = False
+                list(spam.iterator) # Implicitly call ourselves
```

Comment:
This evening (after getting access to my dev machine) I'll check whether the inner `list()` is required to throw `StopIteration` to trigger the bug. If no, I will simplify the test:

```suggestion
                try:
                    list(spam.iterator) # Implicitly call ourselves.
                except StopIteration:
                    pass
```

Reply:
I confirmed that `SystemError` on the top level is thrown both with and without propagation of `StopIteration` from the bottom level. So I'm applying the change.

Reply:
I got it wrong. `list()` cannot throw an exception because it returns a ready-to-use list (possibly empty) processing `StopIteration` by itself.